Modifications to `std::io::Stdin` on Windows so that there is no longer a 4-byte buffer minimum in read().

Patrick-Poitras opened the pull request:
This is an attempted fix of issue #91722, where a too-small buffer was passed to the read function of stdio on Windows. This caused an error to be returned when `read_to_end` or `read_to_string` were called. Both delegate to `std::io::default_read_to_end`, which creates a buffer that is of length >0, and forwards it to `std::io::Stdin::read()`. The latter method returns an error if the length of the buffer is less than 4, as there might not be enough space to allocate a UTF-16 character. This creates a problem when the buffer length is in `0 < N < 4`, causing the bug. 

The current modification creates an internal buffer, much like the one used for the write functions

I'd also like to acknowledge the help of @agausmann and @hkratz in detecting and isolating the bug, and for suggestions that made the fix possible. 

Couple disclaimers:

- Firstly, I didn't know where to put code to replicate the bug found in the issue. It would probably be wise to add that case to the testing suite, but I'm afraid that I don't know _where_ that test should be added.
- Secondly, the code is fairly fundamental to IO operations, so my fears are that this may cause some undesired side effects ~or performance loss in benchmarks.~ The testing suite runs on my computer, and it does fix the issue noted in #91722.
- Thirdly, I left the "surrogate" field in the Stdin struct, but from a cursory glance, it seems to be serving the same purpose for other functions. Perhaps merging the two would be appropriate.

Finally, this is my first pull request to the rust language, and as such some things may be weird/unidiomatic/plain out bad. If there are any obvious improvements I could do to the code, or any other suggestions, I would appreciate them.

Edit: Closes #91722

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4486998249?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5433225dd189e2ee625e1dcd41ea3af41655a792/checks/4486998249/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* 361 features
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:09
```

</details>

agausmann commented on library/std/src/sys/windows/stdio.rs:
```diff
@@ -29,6 +31,32 @@ struct IncompleteUtf8 {
+            buf[ind] = *val;
+        }
+
+        // Rotate the remaining bytes if not enough remaining space in buffer.
+        if usize::from(self.len) > buf.len() {
+            let new_len = self.len - (buf.len() as u8);
+            let mut new_bytes: [u8; 4] = [0; 4];
+            for (ind, val) in self.bytes[to_write..].iter().enumerate() {
+                new_bytes[ind] = *val;
+            }
+            self.bytes = new_bytes;
+            self.len = new_len;
+        } else {
+            self.len = 0;
+        }
```
This section can probably be simplified using [`[T]::copy_within`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.copy_within)

agausmann replied:
E.g.:
```rust
self.buf.copy_within(to_write.., 0)
```

agausmann commented on library/std/src/sys/windows/stdio.rs:
```diff
@@ -29,6 +31,32 @@ struct IncompleteUtf8 {
+    fn read(&mut self, buf: &mut [u8]) -> usize {
+        // Write to buffer until the buffer is full or we run out of bytes.
+        let to_write = cmp::min(buf.len(), self.len as usize);
+        for (ind, val) in self.bytes[..to_write].iter().enumerate() {
+            buf[ind] = *val;
+        }
```
This can also be simplified using [`copy_from_slice`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.copy_from_slice
):

```rust
buf[..to_write].copy_from_slice(&self.bytes[..to_write])
```

pitdicker commented:
I am not sure if there really is a bug here. Please see https://github.com/rust-lang/rust/pull/58454 for a bit of history.

See this comment as a note for the reviewer, I currently don't have the time to form a real opinion.

Patrick-Poitras commented:
@agausmann I've made the two changes you requested above. Thanks so much! 

On a sidenote, I'm trying to write a test that replicates the behavior, but I think the code is going around any sort of testing by checking if console mode is enabled. This is probably why automated testing didn't catch it before. (They're all probably using pipes, or not console mode)

agausmann commented:
@pitdicker 
> I am not sure if there really is a bug here. 

The current behavior makes `read_to_string` and `read_to_vec` non-portable in a non-intuitive and undocumented way - they have unreliable behavior in Windows console environments and will return an error for very specific forms of input (though not that uncommon) that would be fine on other platforms.

And I agree about what was said in https://github.com/rust-lang/rust/issues/91722#issuecomment-991032742 about windows stdio - it ought to be liberal in what it accepts, specifically it should be able to handle all buffer sizes:
> The `io::Read` implementation for `Stdin` should conform to the behavior specified in the [documentation of the Read trait](https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read). Essentially calling `read()` with a buf of 0 < length < 4 should work and return (at least) `Ok(1)` with one byte of the UTF-8 sequence if EOF has not been reached.

Patrick-Poitras commented:
I agree that we should conform to the specifications of the Read function. 

I've also changed this from a draft to an open pull request, after doing enough manual testing that I've convinced myself that the behavior is correct. I've still been unable to find where in the code this path gets tested, or an alternative method for testing it. 

I've removed the request for a benchmark from the main post. Piped-in IO short circuits the modified of the code so I don't see this being a problem that would bench unfavorably on non-console mode path. The console mode path would only be slowed down if the length is less than 4, which currently returns an error, so I believe that that covers any potential performance decreases.

agausmann commented:
> but I think the code is going around any sort of testing by checking if console mode is enabled

Yeah, this section of code is only used when the I/O device is a handle to a Windows ["console"](https://docs.microsoft.com/en-us/windows/console/consoles). I don't know much about this, but as I understand it, Rust converts the programmer's (UTF-8) sequence of bytes to/from a UTF-16 buffer which is needed by the low-level `ReadConsole`/`WriteConsole` calls.

 It seems like there has been difficulty making automated tests for this in the past - e.g. in #58454, they say they "did some _manual_ testing of reading and writing to console" (emphasis mine).

It may be possible to build a test harness for Windows consoles using [pseudoconsoles](https://docs.microsoft.com/en-us/windows/console/pseudoconsoles)? That probably deserves its own issue, but I think it is worth looking into.

ChrisDenton commented:
> Maybe we can expose a doc(hidden) and unstable helper to get access to the raw stdin to test it...

You can get the Windows console input handle by opening the special device `\\.\CONIN$`. See [CreateFileW#Consoles](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew#consoles)

Patrick-Poitras commented on library/std/src/sys/windows/stdio.rs:
```diff
@@ -221,24 +242,48 @@ impl io::Read for Stdin {
+        // If there are bytes in the incomplete utf-8, start with those.
+        if self.incomplete_utf8.len > 0 {
+            bytes_copied += self.incomplete_utf8.read(buf);
+            buf_len -= bytes_copied;
```
Once I made this change and the one with the self.incomplete_utf8.len, it allowed for further simplification into 

```rust
// If there are bytes in the incomplete utf-8, start with those.
// (No-op if there is nothing in the buffer.)
let mut bytes_copied = self.incomplete_utf8.read(buf);
```

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4526145032?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/30c68c4ab57a209afe4b1e4a3525b183d2edb19a/checks/4526145032/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking addr2line v0.16.0
error: casts cannot be followed by ?
   --> library/std/src/sys/windows/stdio.rs:257:34
    |
257 |             let read_bytes: u8 = utf16_to_utf8(&utf16_buf[..read], &mut self.incomplete_utf8.bytes)? as u8?;
    |
help: try surrounding the expression in parentheses
    |
    |
257 |             let read_bytes: u8 = (utf16_to_utf8(&utf16_buf[..read], &mut self.incomplete_utf8.bytes)? as u8)?;
    |                                  +                                                                         +
error[E0277]: the `?` operator can only be applied to values that implement `Try`
   --> library/std/src/sys/windows/stdio.rs:257:34
    |
    |
257 |             let read_bytes: u8 = utf16_to_utf8(&utf16_buf[..read], &mut self.incomplete_utf8.bytes)? as u8?;
    |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `u8`
    = help: the trait `Try` is not implemented for `u8`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `std` due to 2 previous errors
```

</details>

Mark-Simulacrum commented on library/std/src/sys/windows/stdio.rs:
```diff
@@ -221,24 +242,39 @@ impl io::Read for Stdin {
+        let mut bytes_copied = self.incomplete_utf8.read(buf);
+
+        if bytes_copied == buf.len() {
+            return Ok(bytes_copied);
```
nit: can avoid the return keywords

Mark-Simulacrum commented:
Please squash commits, and one more nit -- then we can move ahead.

> > Maybe we can expose a doc(hidden) and unstable helper to get access to the raw stdin to test it...
> 
> You can get the Windows console input handle by opening the special device `\\.\CONIN$`. See [CreateFileW#Consoles](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew#consoles)

This'll be a good start to getting the stdin for the child process right, thanks!

We will need the raw wrapper provided by std, in a UI test or similar, which means externally from std so need something ~public. I can help out directly with plumbing the bits of std out if that would be helpful, or provide some additional instructions beyond what I previously said.

Patrick-Poitras commented:
@Mark-Simulacrum All done!

Mark-Simulacrum commented:
Alright. I'll try to throw together a UI test skeleton and push it up in the next few days, though I don't have easy access to a windows machine to test on right now.

Thanks!

Mark-Simulacrum commented:
@ChrisDenton hm, is there a way to write to that windows console input so we can read to it from the child? My rough assumption is that this isn't necessarily the same as regular stdin for processes spawned through std::process, but maybe I'm wrong about that?

It's fairly easy to imagine exposing the read impl so we can directly test this, but to make that work well we'd need to manufacture a console handle with information to read from, which I'm not sure how to do.

FWIW, I'd be okay moving ahead here if we could verify manually that this works, even if a UI test is difficult to pull together.

ChrisDenton commented:
Hm... well you can use [WriteConsoleInput](https://docs.microsoft.com/en-us/windows/console/writeconsoleinput) but it's a fairy involved API (you have to create each key event). It's also soft deprecated, but it should continue to work on a local console that's not in a restricted security context.

Patrick-Poitras commented:
> FWIW, I'd be okay moving ahead here if we could verify manually that this works, even if a UI test is difficult to pull together.

I`ve also had a really hard time trying to make a test.

For the manual testing, I have manually confirmed that this works with repeated enters, random keyboard typing, weird characters in UTF-8 and UTF-16, and a series of random non-renderable unicode characters. All of the tests showed 100% byte-per-byte matching the input. No crashing/panics was observed, even with egregious nonsense being done.

Mark-Simulacrum commented:
@Patrick-Poitras Is that testing the raw API being modified here? Or the regular io::stdin, which buffers it with BufReader?

> Hm... well you can use WriteConsoleInput but it's a fairy involved API (you have to create each key event). It's also soft deprecated, but it should continue to work on a local console that's not in a restricted security context.

Hm. Yeah, OK. Let's try to confirm functionality through manual testing of this, but not try to write a runnable test, then.

Patrick-Poitras commented:
I did testing in-situ, using console mode in powershell and cmd. It was that way that the bug was found, so explicitly calling it through console mode. Another possibility for testing would be to write a flag that disables the console mode workaround, idk if that's permitted.

Patrick-Poitras commented:
@Mark-Simulacrum Do you need me to do anything else before we get this rolling?

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit d49d1d44995a75fd288f860c10b2bc18cfc8a5b4 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum d49d1d44995a75fd288f860c10b2bc18cfc8a5b4 -->
<!-- homu: {"type":"Approved","sha":"d49d1d44995a75fd288f860c10b2bc18cfc8a5b4","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit d49d1d44995a75fd288f860c10b2bc18cfc8a5b4 with merge 007f52dba2161c8de49b0905c05ecefde0448e21...
<!-- homu: {"type":"BuildStarted","head_sha":"d49d1d44995a75fd288f860c10b2bc18cfc8a5b4","merge_sha":"007f52dba2161c8de49b0905c05ecefde0448e21"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/4682620040?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/007f52dba2161c8de49b0905c05ecefde0448e21/checks/4682620040/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4682885139?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/4682885139?check_suite_focus=true","builder_name":"checks-actions"} -->

matthiaskrgr commented:
@bors retry apple runner did not start...?

bors commented:
:hourglass: Testing commit d49d1d44995a75fd288f860c10b2bc18cfc8a5b4 with merge 51288ba26d420cbfbfc8f9382fb5fe1e45752167...
<!-- homu: {"type":"BuildStarted","head_sha":"d49d1d44995a75fd288f860c10b2bc18cfc8a5b4","merge_sha":"51288ba26d420cbfbfc8f9382fb5fe1e45752167"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

Patrick-Poitras commented:
@matthiaskrgr Same runner on apple timed out. I checked the output while it was running and it was the last build remaining about 3 hours ago. The apple build is running significantly slower than other builds for no apparent reason. (Also this change is Windows-only, so unless something weird is happening, I don't think it's got anything to do with this PR)

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/4686514742?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/51288ba26d420cbfbfc8f9382fb5fe1e45752167/checks/4686514742/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

lqd commented:
@bors retry
